azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +62,166 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
-        if full_goal_state:
-            fetch_full_goal_state = True
-            reason = 'force update'
-        elif base_incarnation is not None and self.incarnation != base_incarnation:
-            fetch_full_goal_state = True
-            reason = 'new incarnation'
-
-        if not fetch_full_goal_state:
-            self.hosting_env = None
-            self.shared_conf = None
-            self.certs = None
-            self.ext_conf = None
-            self.remote_access = None
-            return
-
-        logger.info('Fetching new goal state [incarnation {0} ({1})]', self.incarnation, reason)
-
-        try:
-            uri = findtext(xml_doc, "HostingEnvironmentConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-            self.hosting_env = HostingEnv(xml_text)
+        # Nothing retrieves certificates, so we need to keep this logic the same for now
+        # Limit retrieving certificates to only once per incarnation
+        uri = findtext(self._xml_doc, "Certificates")
+        if uri is None:
+            GoalState._Certs = None
+        elif GoalState._IncarnationForCerts is None or GoalState._IncarnationForCerts != self.incarnation:
+            GoalState._IncarnationForCerts = self.incarnation
+            xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
+            GoalState._Certs = Certificates(xml_text)
 
-            uri = findtext(xml_doc, "SharedConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-            self.shared_conf = SharedConfig(xml_text)
+        self.certs = GoalState._Certs
 
-            uri = findtext(xml_doc, "Certificates")
-            if uri is None:
-                self.certs = None
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
-                self.certs = Certificates(xml_text)
-
-            uri = findtext(xml_doc, "ExtensionsConfig")
-            if uri is None:
-                self.ext_conf = ExtensionsConfig(None)
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-                self.ext_conf = ExtensionsConfig(xml_text)
+        GoalState.ContainerID = self.container_id
 
-            uri = findtext(container, "RemoteAccessInfo")
-            if uri is None:
-                self.remote_access = None
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
-                self.remote_access = RemoteAccess(xml_text)
-        except Exception as e:
-            logger.warn("Fetching the goal state failed: {0}", ustr(e))
-            raise
-        finally:
-            logger.info('Fetch goal state completed')
+        # The changed property on GoalState indicates only if the incarnation changed
+        # There's also a changed property on ext_config that indicates if they changed
+        self.changed = False
```

Comment:
Added a comment below in wire.py, basically I found only 1 use for this property which can also be replaced by ext_conf.changed property. If that's the case please remove this as it feels redundant and unnecessary.

Reply:
see the response above